src/shm/pool/mod.rs

```diff
@@ -27,3 +28,15 @@ impl From<Errno> for CreatePoolError {
         Into::<io::Error>::into(errno).into()
     }
 }
+
+pub trait AsPool<P> {
+    fn pool_handle(&self) -> PoolHandle<P>;
```

Comment:
Let me think about this.

Reply:
I don't think this is needed anymore if I do this
```rust
impl<D, K: PartialEq + Clone> DelegateDispatch<wl_buffer::WlBuffer, D> for MultiPool<K>
where
    D: Dispatch<wl_buffer::WlBuffer, UserData = Self::UserData>,
    for<'p> &'p mut D: Into<PoolHandle<'p, MultiPool<K>>>
{
...
```

Reply:
An other option is to use the `AsMut` trait, that might better reflect the intent

Reply:
The issue with AsMut is that the return type must be a reference and I don't think the reference would live long enough or there would be a mismatch which means lifetime hell.

Reply:
Let's consider an alternative for a bit before we decide which trait flavor is best.

The reason for this trait requirement is looking up the pool at runtime in order to modify internal state. Could we store a inner type of the pool `Arc/Weak<Mutex<InnerMultiPool>>` inside the user data and avoid the need for these lookup functions?

Reply:
I'd rather not use a reference counted pointer when it can be solved with a normal reference. It could also limit how the user wants to organize their pools.

Reply:
As much as I agree with you about not wanting yet more reference counted pointers, wayland is a shared access mess sadly.

Considering this pool is arguably a mid-level pool (a semi double buffering pool needs to be built above this at some point), needing to expose a trait that returns a slice of MultiPool may not exactly work with the higher level abstractions. The shared pointer isn't ideal but it would only ever be internal to the wl_buffer and the pool.

I would think the destruction logic would be the following for buffers:

If the pool is first, destroy all wl_buffers that are not active (or at least queue it up since that needs &mut ConnectionHandle).
When active pools are sent the release event, if the pool has been dropped, destroy the wl_buffer immediately. The reference count hitting zero will finally destroy the inner pool.

Reply:
I am not sure I understand why it would be preferable to have an inner type. What would the InnerMultiPool look like?

I imagined a pool built on top of this one would own it so it wouldn't need to go use `PoolHandle`. 

> needing to expose a trait that returns a slice of MultiPool may not exactly work with the higher level abstractions. 

It returns a `PoolHandle` which is an enum. We could add the inner type to it. We'd still have to use Into to convert the Dispatch but it leaves a lot more option open.